mri/reconstructors/calibrationless.py

```diff
@@ -0,0 +1,108 @@
+# -*- coding: utf-8 -*-
+##########################################################################
+# pySAP - Copyright (C) CEA, 2017 - 2018
+# Distributed under the terms of the CeCILL-B license, as published by
+# the CEA-CNRS-INRIA. Refer to the LICENSE file or to
+# http://www.cecill.info/licences/Licence_CeCILL-B_V1-en.html
+# for details.
+##########################################################################
+
+"""
+This implements calibrationless reconstruction with different proximities
+"""
+# System import
+import warnings
+
+from .base import ReconstructorBase
+from ..operators import GradAnalysis, GradSynthesis, WaveletN
+
+# Third party import
+from modopt.opt.proximity import SparseThreshold
+from modopt.opt.linear import Identity
+
+
+class SparseCalibrationlessReconstructor(ReconstructorBase):
+    """ This class implements a calibrationless reconstruction based on the
+    L1-norm regularization.
+    For the Analysis case finds the solution for x of:
+        (1/2) * sum(||F x_l - yl||^2_2, n_coils) +
+                    mu * sum(||Wt x_l||_1, n_coils)
+    For the Synthesis case finds the solution of:
+        (1/2) * sum(||F Wt alpha_l - yl||^2_2, n_coils) +
+                    mu * sum(||alpha_l||_1, n_coils)
+    Parameters
+    ----------
+    fourier_op: object of class FFT, NonCartesianFFT or Stacked3DNFFT in
+                mri.operators
+        Defines the fourier operator F in the above equation.
+    linear_op: object, (optional, default None)
+        Defines the linear sparsifying operator Wt. This must operate on x and
+        have 2 functions, op(x) and adj_op(coeff) which implements the
+        operator and adjoint operator. For wavelets, this can be object of
+        class WaveletN or WaveletUD2 from mri.operators .
+        If None, sym8 wavelet with nb_scales=3 is chosen.
+    prox_op: operator, (optional default None)
+        Defines the proximity operator for the regularization function.
+        For example, for L1 Norm, the proximity operator is Thresholding
+        If None, the proximity opertaor is defined as Soft thresholding
+        of wavelet coefficients with mu value as specified.
+    mu: float or np.ndarray, (optional, default 0)
```

Comment:
`mu` should  be removed since the user provide the proximity operator the mu value doesn't make any sense here.
If you just want a value put it to 0

Reply:
It is zero by default, and if user does not provide a prox_op, we just place sparsity in wavelet coefficients and obtain the results. We could make some modifications and implement groupLasso as default if needed. 
The point is, this would take away burden from user from specifying proximity always, we have a default proximity ready at hand.

Reply:
This is very specific to the L1-norm, other norms have more than one hyper-paremeters to set, this is also the case for the group-LASSO, therefore there is no point of adding.
In my opinion if any proximity operator have been provided then we should only consider a gradient descent algorithm and set the proximity operator to Identity, You will just need to inform the user through a warning

Reply:
well in that case, I will agree to remove this parameter only for calibrationless as a pure L1 norm reguralizer kinda doesnt make sense. 
Adding here that the warning already exists:
```python
if prox_op is None and mu == 0:
            warnings.warn("The prox_op is not set and mu = 0. The result will "
                          "not be a reconstruction but an inverse")
```

So we agree to remove this parameter only for Calibrationless or do we just keep everything as is with the warning as above which is a general warning for all reconstructors (resides in base)

Reply:
Just to be clear the inverse Fourier transform is already a reconstruction the warning here is incorect.
If  `prox_op=None` then the proximity should be equal to `Identity` and the mu should be removed

Reply:
This is updated anyway as we removed mu, resolving.